0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+This means extra cognitive load for Dafny users, extra implementation effort for any new builtin collection types, 
+and additional testing to ensure that all supported quantification domains are deterministically enumerable.
+
+It can be argued that if any compiled Dafny program uses existing features that depend on domain enumeration,
+such as `exists` expressions or `:|` statements, its runtime behavior already depends heavily on how this domain
+searching is implemented. The compiler only enforces that such domains are finite and enumerable, but an inefficient
+choice can lead to the search for a matching value taking orders of magnitude more time than expected.
+Therefore we should already be documenting this aspect of Dafny semantics and implementation better anyway.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternative is to only provide the simpler `foreach x in C` syntax, where `C` is any collection value.
+This is certainly a well-formed and sound construct, but far less powerful than the proposed version, especially
+when sequence comprehensions are excluded. It also still runs into some of the same semantic challenges around
+ordering if `C` is allowed to be a set or multiset anyway, since verification of such a loop still needs to be aware of
+this ordering. Offering `foreach` loops along with sequence comprehensions means we can recommend the latter where possible,
+with the former as a fallback that is inevitably more complicated to verify.
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`
+
+Most mainstream programming languages include rich libraries for working with collections and iterators,
+and define much the same operations such as "filter" and "flatMap".
+The proposed Dafny features provide similar functionality at a higher level
+of abstraction more amenable to verification: quantification expressions are used to succinctly declare the behavior of
+an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
+pushing this philosophy even further.
+
+The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
+pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
+a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
+sums the indices of all non-zero elements of the array `a`. The JMatch version declares the bound variables inline in
+the boolean expression, which is also proposed here as a future possibility below.
+
```

Comment:
There's another instance I'm aware of that provides a capability not available in this proposal. I don't know if it's worth the complexity of supporting, but perhaps worth considering (or perhaps including in here to say why it's not the right thing to do).

[Cryptol](https://cryptol.net/) has two sequence comprehension separators, sequential and parallel. The sequential separators work like those in Python (and Haskell), and like `&&` in this proposal. For example, the cross product of two sequences is:

`crossProduct xs ys = [ (x, y) | x <- xs, y <- ys ]`

But you can also use a `|` instead of a comma to get this:

`zip xs ys = [ (x, y) | x <- xs | y <- ys ]`

You can also get this by zipping two iterators together first and then iterating over the result, which isn't much harder. The slightly more powerful (but sometimes confusing!) use case is when you alternate `,` and `|` in arbitrary ways.

Reply:
Interesting! Cryptol's version is at least worth naming as prior art as it's even closer to this proposal - thanks for the pointer!

My plan is more likely to handle that case when we support user-defined collection types. I included zipping as a key higher-order operation in [these experiments](https://github.com/dafny-lang/libraries/pull/37/files#diff-165e6a6e1f52e2d36e8c33448d2a99bc1553b21457565d3995247ed626a3e4faR490), and I'm fairly convinced it's maximally "Dafny-like" to require the LHS and RHS have provably-equal lengths. Between that and the fact that there's no natural boolean operator to express zipping, I'd say it's it doesn't fit as well here.